### ISSUE TITLE
Optional prompting with vendoring leading to blank page

### ISSUE THREAD

**Issue by adrianrudnik** [antfu/vite-plugin-pwa#177]
Hi there, sorry for this messy question. This is the second time I try to integrate this plugin and get stuck with a blank page and this time I want to get to the bottom of it. I've look through most of #33 but my problem seems to be only slightly related.

I have a demo site that requires a promp to avoid users loosing an active voice recording session, so they can click "do it later" on new releases.

The setup is a simple vue + vite setup, default `VitePWA` settings taken from [the intro](https://vite-plugin-pwa.netlify.app/guide/generate.html). I build it with `npm run build` and have a `index.html` as entry point, leading now to i.e. `/assets/index.901aa92b.js`. No SSG, no SSR.

I release a new version by copying the generated, static, files into a new nginx docker image to serve them. The service worker was registered correctly before, the prompt works if I click on "reload now", if I click on "later/close" I sometimes end up in blank site hell.

![2021-11-29_00-01](https://user-images.githubusercontent.com/681514/143790037-9cee7deb-020c-4b98-bcd3-341956e89782.png)

It looks lile the service worker lost its own cache, but it does not seem size relevant as mentioned in [the docs](https://vite-plugin-pwa.netlify.app/guide/faq.html#missing-assets-from-sw-precache-manifest). The `index.html` is cached, but not refreshed when the vendor breaks. The site itself is not working against the storage limit:

![2021-11-29_00-13](https://user-images.githubusercontent.com/681514/143790237-acd3814b-fb5c-479d-89fe-66e15ba96af8.png)

I know that the old vendor file `index.c3cdb38f.js` that is now missing, was part of the old version but can no longer be loaded as it was prebuild and served by the old docker image.

The old service worker is still alive and the new one is in the "waiting to active" state:

![2021-11-29_00-21](https://user-images.githubusercontent.com/681514/143790487-de09cd01-ae0b-4292-8f2f-219523f16c3a.png)

The `ReloadPrompt.vue` from [the example](https://vite-plugin-pwa.netlify.app/frameworks/vue.html#prompt-for-update) is now in a vendor file no longer available to the client. Now I hope that this is the right lead, as I ran out of ideas. What is the correct approach here?

Should I avoid using clean/focused images and instead deploy in a incremental matter (like ftp/scping all new files into the existing folder structure)?

As far as I understood the `networkFirst` strategy from #33 would also fail in this scenario, right?

Did I miss anything crucial from the manual?

**Comment by userquin** [comment]
I need the version of the plugin you are using and the pwa configuration, tmr Ill take a look...

**Comment by userquin** [comment]
@adrianrudnik I think it is something related with the assets cache. Once the new sw is on `waiting` state, the current assets are old ones, so it should work (stale while revalidate). Of course if the user do a hard refresh (without reloading the sw), then you will get a blank page.

The only assets that should be server with no cache should be ìndex.html` and `sw.js`, any other should be server with immutable, see for example `netlify.toml` and `public/_headers`  on `vitesse template` or the same resources here on docs package/directory.

**Comment by userquin** [comment]
I have some projects using `http/2` and are working without any problem (only when hard refresh, with f5 works again).

**Comment by userquin** [comment]
For example, using netty server with spring webflux with http/2, all assets directory are server with `ETag` and `immutable` since each build will generate a new version (I generate the `manifest` from the deps on each page):
<details>
<summary>ETag</summary>

```kotlin
    override fun addResourceHandlers(registry: ResourceHandlerRegistry) {
        val resources = webProperties.resources
        registry.addResourceHandler(
            "/**"
        ).addResourceLocations(
            *resources.staticLocations
        ).setCacheControl(
            resources.cache.cachecontrol.toHttpCacheControl()
        ).resourceChain(
            resources.chain.isCache
        ).addTransformer { exchange, resource, _ ->
            Mono.fromSupplier {
                resource.lastModified()
            }.flatMap { lastModified ->
                exchange.response.headers.apply {
                    addHeaders(lastModified, false)
                    // ETag
                    resource.filename?.let { name ->
                        when {
                            name.endsWith(".js") || name.endsWith((".css")) -> {
                                name.substring(0, name.lastIndexOf('.')).let { nNoExt ->
                                    val index = nNoExt.lastIndexOf('.')
                                    if (index > 0) {
                                        this.eTag = "W/\"${nNoExt.substring(index + 1)}\""
                                    }
                                }
                            }
                            name.startsWith("workbox-") && name.endsWith(".js") -> {
                                name.substring(0, name.lastIndexOf('.')).let { nNoExt ->
                                    val index = nNoExt.lastIndexOf('-')
                                    if (index > 0) {
                                        this.eTag = "W/\"${nNoExt.substring(index + 1)}\""
                                    }
                                }
                            }
                            else -> {}
                        }
                    }
                }
                Mono.just(resource)
            }
        }/*.addResolver(
            resourceFallbackResolver
        )*/.addResolver(
            EncodedResourceResolver()
        ).addResolver(
            PathResourceResolver()
        )
    }
```
</details>

and `addHeaders` is:

<details>
<summary>addHeaders</summary>

```kotlin
    fun HttpHeaders.addHeaders(lastModified: Long, mainPage: Boolean) {
        val now = Instant.now().toEpochMilli().apply {
            add("age", "0")
        }
        date = now
        expires = now
        this.lastModified = lastModified
        add("referrer-policy", "no-referrer")
        add("strict-transport-security", "max-age=31536000")
        add("x-content-type-options", "nosniff")
        .... // some other logic
    }
```
</details>

**Comment by adrianrudnik** [comment]
@userquin Thank for the feedback! The version used is 0.11.8, the pwa configuration looks like this:

```
    VitePWA({
      registerType: 'prompt',
      includeAssets: ['favicon.svg', 'favicon.ico', 'robots.txt', 'apple-touch-icon.png'],
      manifest: {
        name: 'xxx',
        description: 'xxx',
        short_name: 'xxx',
        theme_color: '#ffffff',
        background_color: '#000000',
        display: 'standalone',
        lang: 'en',
        orientation: 'any',
        start_url: '/?standalone=true',
        scope: '/',
        icons: [
          {
            src: 'pwa-192x192.png',
            sizes: '192x192',
            type: 'image/png',
          },
          {
            src: 'pwa-512x512.png',
            sizes: '512x512',
            type: 'image/png',
          },
          {
            src: 'pwa-512x512.png',
            sizes: '512x512',
            type: 'image/png',
            purpose: 'any maskable',
          },
        ],
      },
    }),
```

> stale while revalidate

I think this could be the problem. They might be stale, but during revalidation the old asset is 404 on the server, so the cache might become valid if it is no longer present on the server?

Just to make sure: The image I build with docker only contains the new build files, not the old ones that get replaced, so they will be missing.

> The only assets that should be server with no cache should be index.html and sw.js

Can this work when I use the prompt method without automatic installation? The prompt (a Vue SFC) itself will get compiled into a vendor file. I'm not using SSG, so index.html is pretty minimal, only using script tags to load everything else. I just did a grep through the vendor files and the update prompt is indeed in the `assets/index.11f8bda0.js` (at least for now), so is every other "boot" file required for the site to show.

How do you deploy, do the old vendor files get removed or do you keep them in place?

**Comment by adrianrudnik** [comment]
> Of course if the user do a hard refresh (without reloading the sw), then you will get a blank page.

That sounds exactly like the problem. I guess the only solution would be to keep the old vendor files, so they can be reloaded by a stale, but installed, service worker. Every other scenario actually requires me to uninstall the service worker, no amount of hard refreshes help to get rid of the blank page.

Is there any help or doc site to get an overview of the recommended cache-control headers for sw.js, vendor files and the related pwa assets?

**Comment by userquin** [comment]
To keep old assets on the sw you can configure `cleanupOutdatedCaches: false` (beware disabling that option). 

I always delete the old dist folder from server and also enable `cleanupOutdatedCaches: true` to remove the old assets once the new sw activated.

You can also keep old assets on the server, this will prevent the error while sw not updated and hard refresh from client.

About cache control, I use ETag (vite generates the assets with hash, I use that hash to build the ETag).

`ReloadPrompt` also works, I use this behavior 99%. Beware when you switch from `autoUpdate` to `prompt` on production, since it will be only shown when the user install the new sw 2 times, the first one will still use `autoUpdate`, once installed, next new version will show the prompt.

You can test your application on [WebPageTest](https://www.webpagetest.org/), it should hint you about the cache on your application.

About the docs for cache search on google, there are a lot of good stuff out there.

**Comment by adrianrudnik** [comment]
Thanks! I will adapt an copy-over-old approach for now, as this avoids the problem at its core.

But I will also take away some review for the headers as well. I just did a review of the NGINX vhost configuration used for the site, which is a pretty basic 0815 template for this stuff, same I used for precompile hugo sites with static assets with hashes:

```
    location ~* \.(?:ico|css|js)$ {
        access_log off;

        add_header Cache-Control "public,max-age=31536000,s-maxage=31536000,immutable" always;

        add_header Vary Accept;
        try_files $uri =404;
    }
```

The error is in the detail that `.js` files get long cache times and the immutable flag. It is applied to the `sw.js` as well but must not right? I cant find any documentation about the behavior on requests serving etag and immutable.

Thanks @userquin , maybe I will do some pull request with updated details on the used nginx configuration if that would improve anything.
